core/trino-main/src/main/java/io/trino/operator/TopNProcessor.java

```diff
@@ -46,28 +48,33 @@ public TopNProcessor(
             List<Type> types,
             int n,
             List<Integer> sortChannels,
-            List<SortOrder> sortOrders, TypeOperators typeOperators)
+            List<SortOrder> sortOrders,
+            TypeOperators typeOperators)
     {
         requireNonNull(aggregatedMemoryContext, "aggregatedMemoryContext is null");
-        this.localUserMemoryContext = aggregatedMemoryContext.newLocalMemoryContext(TopNProcessor.class.getSimpleName());
+        this.localMemoryContext = aggregatedMemoryContext.newLocalMemoryContext(TopNProcessor.class.getSimpleName());
         checkArgument(n >= 0, "n must be positive");
 
         if (n == 0) {
-            outputIterator = emptyIterator();
+            topNBuilderSupplier = null;
         }
         else {
-            topNBuilder = new GroupedTopNRowNumberBuilder(
+            GroupByHash noChannelGroupByHash = new NoChannelGroupByHash();
+            PageWithPositionComparator comparator = new SimplePageWithPositionComparator(types, sortChannels, sortOrders, typeOperators);
+            topNBuilderSupplier = () -> new GroupedTopNRowNumberBuilder(
                     types,
-                    new SimplePageWithPositionComparator(types, sortChannels, sortOrders, typeOperators),
+                    comparator,
                     n,
                     false,
-                    new NoChannelGroupByHash());
+                    noChannelGroupByHash);
         }
     }
 
     public void addInput(Page page)
     {
-        requireNonNull(topNBuilder, "topNBuilder is null");
+        if (topNBuilder == null) {
+            topNBuilder = topNBuilderSupplier.get();
```

Comment:
add `requireNonNull(topNBuilderSupplier, "topNBuilderSupplier is null");`

Reply:
Done